[FEATURE] Add new rendering theme

benjaminkott opened the pull request:
### How to Test:

1. Install 
    ```
    yarn install
    yarn build
    ```
    This will render the "How to Document Guide" as reference.
    If you want to test a different documentation, simply provide the path.
    ```
    # full build
    yarn build --source=..\t3SphinxThemeRtdDemoDocs

    # just render
    yarn render --source=..\t3SphinxThemeRtdDemoDocs
    ```

2. Start HTTP-Server
    ```
    yarn serve
    ```
3. Validate
    Open Browser on the Served Address and check if everything is working

### Testing

Each change to the theme will trigger a GitHub action build, that will provide an artifact of a rendered documentation of https://github.com/TYPO3-Documentation/TYPO3CMS-Guide-HowToDocument. See: https://github.com/benjaminkott/typo3-sphinx-theme/actions

1. 
![image](https://user-images.githubusercontent.com/3243119/71807915-b933c380-306c-11ea-9af3-814659241d88.png)
2. 
![image](https://user-images.githubusercontent.com/3243119/71807960-d8325580-306c-11ea-9ee4-58a1d268d972.png)



### Fixes: 

- **#122 Contrast for some syntax highlighting colors - readability could be improved** 
  A new color theme is in place.

- **#110 Color scheme (orange) of docs.typo3.org inconsistent with TYPO3 style guide**
  Streamlined for now, will be discussed with design team.

- **#109 Using a sidebar at end of page breaks rendering**
  Sidebar does not break the rendering anymore.

- **#104 Improve readability by adding more space around images**
  Image display is now streamlined with margins.

- **#103 Make toolchain and documentation for toolchain "contributor-ready"**
  **#85 Building t3SphinxThemeRtd fails with 404**
  Theme development is now reduced to the following steps, it will always take the current installed documentation docker container or build pull a new one.
  ```
  yarn install
  yarn build
  yarn serve
  ```
- **#92 Changed urls for links in contentinfo section (bottom)**
  Links are now updated.

- **#94 Add homepage link to breadcrumb and change anchor text for start of manual**
  Breadcrumb root node was renamed to `Start`

- **#89 Dimmed buttons should be better visible - a bit**
  Pager on top was removed.

- **#81 Restyle and restructure left navigation panel**
  Navigation panel was rebuild from the ground up.

- **#141 Next / prev page causes new page to "jump" while rendered**
  Introduced a hidden navigation on top that is not displayed and not in the tabindex,
  this navigation is only used to provide targets for previous and next access key navigation.
  This prevents jumping to the bottom navigation when using the shortcuts.

- **#133 Position "Related links" correctly**
  Related Links bar was reworked and does not exist anymore in its current form.
  The version selector and contents have been moved to different locations.

- **#118 Release version is not displayed on mobile**
  Version/Release is not hidden anymore on mobile and always visible.

- **#64 Color change for :hover not working for visited links**
  Links are now changing their hover color correctly.

- **#138 Long inline code will be displayed without line breaking and ends up outside of viewport**
  Long inline code lines are now breaking into the next row when they are too long.

benjaminkott committed:
[TASK] Update dependencies

benjaminkott committed:
[REBUILD] WIP

benjaminkott committed:
[TASK] Use t3docs/render-documentation

benjaminkott committed:
[TASK] Use serve instead of connect and open

benjaminkott committed:
[TASK] Update and remove unnessesary dependencies

benjaminkott committed:
[TASK] Minify javascript files

benjaminkott committed:
[TASK] Remove sourcemap for css files

benjaminkott committed:
[TASK] Extract webfonts

benjaminkott committed:
[TASK] Cleanup indention

benjaminkott committed:
[TASK] Code styling

benjaminkott committed:
[TASK] Add fontawesome

benjaminkott committed:
[TASK] Use placeholder images

benjaminkott committed:
[WIP] Migrate styles

benjaminkott committed:
[WIP] Add more styling

benjaminkott commented:
FYI: @marble @DanielSiepmann @sypets I could start working early on the new theme, added a draft pr so you can follow the development. As soon as I reach a "review" worthy state, I´ll let you know.

benjaminkott committed:
[BUGFIX] Correct overflow issues

benjaminkott committed:
[TASK] Add textroles

benjaminkott committed:
[TASK] Adjust typesetting

benjaminkott committed:
[TASK] Update py module

benjaminkott committed:
[TASK] Add math and parameter args

benjaminkott committed:
[TASK] Cleanup files and migrate add utilities

benjaminkott committed:
[TASK] Remove unnessesary code

benjaminkott committed:
[TASK] Migrate tablerow and adjust colors and font sizes

benjaminkott committed:
[TASK] Add literals and cleanup

benjaminkott committed:
[TASK] Streamline casing of demo file names

benjaminkott committed:
[TASK] Add keystrokes and guilabel

benjaminkott committed:
[BUGFIX] Correct table wrapping

marble commented:
@benjaminkott Fine, thank you, I'll keep watching this and give it a - local - try. cc @DanielSiepmann cc @sypets

benjaminkott committed:
[TASK] Add settings cfg to documentation demo

benjaminkott committed:
[TASK] Bundle fontawesome brands

benjaminkott committed:
[TASK] Add simple workflow

benjaminkott committed:
[BUGFIX] Correct syntax error in workflow file

benjaminkott commented:
@marble added a simple workflow that is rendering the demo docs included in this repository. You can check them currently in my branch: 

https://github.com/benjaminkott/typo3-sphinx-theme/commits/task/add-universe-bar

See Example run: 
https://github.com/benjaminkott/typo3-sphinx-theme/commit/7d3fa6308bd9b8932e6b6c410959e69fbd32cb3b/checks?check_suite_id=348642289

![image](https://user-images.githubusercontent.com/3243119/70422272-bfce1b80-1a6b-11ea-817b-a46f7d58f0cc.png)

cc: @DanielSiepmann @sypets

benjaminkott committed:
[TASK] Add classes to github button

benjaminkott committed:
[TASK] Small adjustments

benjaminkott committed:
[TASK] Use TYPO3CMS-Guide-HowToDocument to test rendering

benjaminkott committed:
[TASK] Also build on pull request

benjaminkott committed:
[TASK] Migrate horizontal buttons

benjaminkott committed:
[TASK] Migrate dl-parameters

benjaminkott committed:
[TASK] Migrate blockquote

benjaminkott committed:
[TASK] Remove unused files

benjaminkott committed:
[WIP] Menu styling

benjaminkott committed:
[TASK] Move show source bar

benjaminkott committed:
[TASK] Show pager only on bottom

benjaminkott committed:
[TASK] Add expandable navigation

benjaminkott committed:
[TASK] Improve menu rendering

benjaminkott committed:
[TASK] Improve navigation bar

benjaminkott committed:
[TASK] Use http-server instead of serve

benjaminkott committed:
[TASK] Layout and table improvements

benjaminkott committed:
[TASK] Always load search and autocomplete

benjaminkott committed:
[TASK] Add frame component

benjaminkott committed:
[TASK] Add panel

benjaminkott committed:
[TASK] Minor styling adjustments

benjaminkott committed:
[TASK] Cleanup

benjaminkott committed:
[TASK] Cleanup

benjaminkott committed:
[TASK] Make standalone

benjaminkott committed:
[BUGFIX] Add missing stemmer and repair search

benjaminkott committed:
[BUGFIX] Add access keys again

benjaminkott committed:
[TASK] Merge breadcrumb with edit and source

benjaminkott committed:
[TASK] Make menu toc basic collapsible

benjaminkott committed:
[TASK] Update footer urls

benjaminkott committed:
[TASK] Rename breadcrumb root to start

benjaminkott committed:
[TASK] Migrate version selector

benjaminkott committed:
[FEATURE] Make source directory for docs configurable

benjaminkott committed:
[TASK] Cleanup leftovers

benjaminkott committed:
[TASK] More leftovers

gilbertsoft commented:
Looks great! Nice job Benji!

benjaminkott committed:
[TASK] Avoid jumping when using access keys - fixes #141

marble commented on Gruntfile.js:
```diff
@@ -1,136 +1,261 @@
     exec: {
       build_sphinx: {
-        cmd: './my-sphinx-build.sh demo_docs/source demo_docs/build'
+        command: 'docker run --rm '
```
On Linux Docker creates files that belong to the `root` user. So only the first build succeeds, because the user is not allowed to clean those file. (S)he needs to `sudo chown -R $(id -u):$(id -g) build/` between builds. For Linux users the correct line would be `command: 'docker run --rm --user=$(id -u):$(id -g) '` This is fine with Macs too, but doesn't work for Windows users. I don't know if Windows users are using this `yarn` and `grunt` stuff. If we have to account for Windows users, I'd leave the line as it is. Otherwise I recommend to add the `--user=$(id -u):$(id -g)` option.

benjaminkott replied:
I am Windows User 😂

benjaminkott replied:
@marble please check the bugfix I provided, not sure of we also need to set the user rights on macs too. https://github.com/TYPO3-Documentation/t3SphinxThemeRtd/pull/140/commits/06986e8ece9ef47e0f3d8f1de1f3bceecb2ac326

marble replied:
AFAIR Macs don't core. But I know that `$(id -u)…` works for them. The thing is, that Windows users don't need that.

benjaminkott replied:
Ok, I´ve added a check for Linux users to add the stuff for now. I will try to find someone with a mac to also check if it's necessary here.

benjaminkott committed:
[BUGFIX] Ensure files are written as current user on linux

benjaminkott committed:
[TASK] Add usercentrics for cookie management

benjaminkott committed:
[BUGFIX] Remove leftovers

benjaminkott committed:
[BUGFIX] Correct syntax error in build jobfile

marble reviewed (commented):
Wow, your are the hero. That's really an amazing work and a jump right into looking "professional". I love it! What I noticed or think about - in random order:
- In `jobfile.json` there should be [1] as only empty strings are counted as FALSE. That's a bit unfortunate, but well, that's how it is at the moment.

- Borders in tables are a bit inconsistent and looking weird (as before):  
![grafik](https://user-images.githubusercontent.com/307057/72147476-e222c400-339e-11ea-889d-47ffbb5abf34.png)

- Homepage: Inline-big-buttons don't behave well in long lines (as before): 
![grafik](https://user-images.githubusercontent.com/307057/72146083-06c96c80-339c-11ea-80ce-141a8a6bf1be.png)

- Link color: I know it's the correct TYPO3 color, but when used for links, it feels too bright for me, very much tiring the eyes: 
![grafik](https://user-images.githubusercontent.com/307057/72146990-e3072600-339d-11ea-8d2b-ae3a62caf6cb.png)
Personally I'm not too upset about this any more as I'm using the "Darkreader" extension in my browsers and these work great with the theme. Otherwise I think I would refuse to read longer texts. But still: Though it's probably the same color, the linktext looks brighter to me than the buttons for example. Personally I would make the link text darker.
 
- I've not tested the version switcher but assume it's working.

- In the `confval` and `sphinxcontrib-phpdomain` directive there is that "blue" that looks a bit strange to me. It is a color of the original Sphinx theme that never had been adapted to the TYPO3 look and feel before: 
![grafik](https://user-images.githubusercontent.com/307057/72147810-b2c08700-339f-11ea-97da-4f9ae3dea5f7.png)
Or here in "this and that": 
![grafik](https://user-images.githubusercontent.com/307057/72147969-00d58a80-33a0-11ea-9e3c-0284fc2348bf.png)




[1]
```{
    "00-INFO": "Use the empty \"\" string for FALSE",
    "Overrides_cfg": {
        "html_theme_options": {
            "add_piwik": "",
            "docstypo3org": "yes",
            "is_homepage": "",
            "show_legalinfo": "yes"
        }
    }
}
```

marble commented:
One more:
How to provide help to the readers?

Readers often complain about the documentation and don't understand that they can contribute themselves. Some don't want to understand that, but - well... So there is the idea to have a link on each page that directs readers to a landing page in "Writing Documentation".

How and where would we put that?

My opinion: I would leave the beginning of the page exactly as it is now. And the button is fine, as we'll soon have there: "Edit on GitHub", "Edit on BitBucket", "Edit on GitLab". For me the footer seems the right place. Maybe we can have a "prominent" link there, that's sticking out a bit. Supported by an icon? cc → @sypets 
![grafik](https://user-images.githubusercontent.com/307057/72148966-5f036d00-33a2-11ea-9a34-de74191c90b6.png)

benjaminkott commented:
@marble thx for checking the rendering. I am aware of the link color but adapted it, for now, to be consistent for most of the platforms. As mentioned I will discuss this with the design team on their sprint in the TYPO3 Office end of January. To have an agreement about the colors and adapt it afterward on all Platforms.

I will check the styling last styling issues later and fix them.

For the Contributing Part, I've found https://github.com/TYPO3-Documentation/t3SphinxThemeRtd/issues/120 that has more or less the same goal. I would take care of this separately in a dedicated patchset.

There were also tickets about adding additional information to the footer. For now, we have a framework we can then extend with this information over time, and it should be more easy to test and contribute then.

benjaminkott committed:
[TASK] Update dependencies

benjaminkott committed:
[BUGFIX] Correct table styling

benjaminkott committed:
[BUGFIX] Correct spacing after linebreaks for inline buttons

benjaminkott committed:
[BUGFIX] Hide focus on paging

benjaminkott committed:
[BUGFIX] Adapt confval and children

benjaminkott committed:
[BUGFIX] Correct inconsistencies in parameter listing

benjaminkott commented:
@marble 

I could fix the tables and inline-buttons, they should now be fine.

For the `conval` i am lacking examples in the "How to Document Guide", i´ve checked the sphinx rendering for reference. I fixed some minor styling issues but i would keep the "blue" for now. It is the color we use to provide information that makes sense for me to also use them here. Also the highlights within the conval are currently matching out code-highlights, I corrected font-size, and coloring of methods, so there is a visual difference in the stack.

![image](https://user-images.githubusercontent.com/3243119/72203860-ab69ae00-3471-11ea-9650-eb5bc001af30.png)

I would personally go with this state now, as long as there are no major bugs anymore.

marble reviewed (changes requested):
Looks like the Javascript for `.. math::` is missing:
![grafik](https://user-images.githubusercontent.com/307057/72626241-a5307180-394a-11ea-8193-6a1cb6ed3c5a.png)

Should be, as on https://docs.typo3.org/m/typo3/demo-t3SphinxThemeRtd/master/en-us/ThisAndThat/Index.html
![grafik](https://user-images.githubusercontent.com/307057/72626381-eb85d080-394a-11ea-977d-d9e198a70cdd.png)

marble reviewed (changes requested):
Tables still look strange:
![128](https://user-images.githubusercontent.com/307057/72632326-57ba0180-3956-11ea-9920-0f4423c62393.png)

There is some css like:
![129](https://user-images.githubusercontent.com/307057/72632360-6accd180-3956-11ea-8e65-9ed9c8a5144d.png)

If I change the `border-top` to `border` I'm getting a nice table like this:
![127](https://user-images.githubusercontent.com/307057/72632432-90f27180-3956-11ea-851d-2a38503a69d6.png)

I would much more like this appearance. I'm not sure about what's needs to be done for that however.

benjaminkott committed:
[BUGFIX] Add math lib

benjaminkott committed:
[TASK] Improve handling of css and js files

benjaminkott commented:
@marble the math ajax got missing along the way while removing unwanted assets- missed that one. I am not unsetting the files directly, means if plugins provide assets they will be loaded.

Still am not sure about the tables. Checked it in Chrome, Edge and Firefox after I applied the changes. 
You are sure you had the latest changes in this pr?

![image](https://user-images.githubusercontent.com/3243119/72638952-dd44ae00-3964-11ea-9a1d-589c0565f479.png)

marble reviewed (changes requested):
Blockquotes:
It would be cool if we could have them styled a bit more. I've created a demo page that can be used for demonstrating: https://docs.typo3.org/m/typo3/demo-t3SphinxThemeRtd/master/en-us/Blockquotes/Index.html
`<blockquote>` should be indented - that's already done. But still it's not very obvious that it's a quote. And the last line, the 'attribution' (for example: `-- Sherlock Holmes`) is a `<p.attribution></p>` that hasn't styling yet. Should get `text-align: right; font-style: italic`
![grafik](https://user-images.githubusercontent.com/307057/72733242-c043f100-3b97-11ea-9737-646f4634049a.png)
About ideas to make the "quote" character visible: see https://github.com/TYPO3-Documentation/t3SphinxThemeRtd/issues/60 and https://1stwebdesigner.com/css-snippets-blockquotes/
I wouldn't go too fancy though.

marble commented:
> You are sure you had the latest changes in this pr?
@benjaminkott You are right, I must have done something wrong, as tables are looking perfect now. Great, it's almost done!

benjaminkott committed:
[TASK] Improve blockquote styling

benjaminkott commented:
Added more styling for blockquotes

![image](https://user-images.githubusercontent.com/3243119/72737628-33e9fc00-3ba0-11ea-9a63-9646399c8518.png)

marble commented:
Looks good. How about one of these characters as indicator?: ” or „ or ❝ or ❞ or ❠ ?

benjaminkott committed:
[TASK] Adapt blockquote indicator

benjaminkott commented:
Sure, why not - adapted the indicator 

![image](https://user-images.githubusercontent.com/3243119/72738940-ef139480-3ba2-11ea-99a3-c0284cdffac0.png)

marble commented:
@benjaminkott Does the 'ajaxversions' call work for you? How do you test that? Example uri would be https://docs.typo3.org/services/ajaxversions.php?url=https://docs.typo3.org/m/typo3/reference-coreapi/master/en-us/

benjaminkott commented:
@marble there are basically 3 ways:

1. I've checked the response and inserted it manually in the markup.

![image](https://user-images.githubusercontent.com/3243119/72850462-71867c00-3ca9-11ea-897f-87b12e918b6a.png)

![image](https://user-images.githubusercontent.com/3243119/72850319-240a0f00-3ca9-11ea-9cec-7cb368ed80a2.png)

2. I've faked the return value during testing and injected the markup directly.

3. I've added a static route during development that was used instead of the production url during testing.

marble commented:
@benjaminkott ① Yes, I did the same and inserted that manually. That's looking alright.  ② Additionally I faked the request which showed a "200 ok" in the devtools network pane but didn't show up on the page. But I didn't investigate that further because I first wanted to ask. ③ BTW: More asking you, for my learning: If we wanted to be really "pingelig", should that be `<dd>No data available.</dd>`. But I agree, that `<p>`tag looks nicer.

Summary: ATM I'm not sure the Ajax-data really shows up

benjaminkott commented:
Did not change it really only moved the position. In general I am also not so happy with the markup provided by the Ajax response. But did not want to touch it in this run. Would opt for opening an issue for after the rollout. Because we need to change the behavior on both ends.
